### PR TITLE
feat: add forkchoice update tracker type

### LINKED ISSUE #2309: Track Safe and Finalized headers
### Describe the feature

This is currently implemented in

https://github.com/paradigmxyz/reth/blob/534d23eed8798807de9bfe6b5f9027bd5c7e6eb0/crates/storage/provider/src/providers/mod.rs#L114-L121

ideally, this should be the last processed FCU 

#2294 would give use a bit more flexibility for this

I'm not sure if deriving the finalized,safe blocknumbers by subtracting epochs from the latest is accurate

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes https://github.com/paradigmxyz/reth/issues/2309

ideally, we store finalized+safe+latest block numbers in memory, the will be accessed mainly from rpc so they should not go into the blockchain tree (guarded via rwlock).

instead, I believe we should add a few individual Arc<Rwlock> here:

https://github.com/paradigmxyz/reth/blob/39218b1ee82b5ee3b39b0e0278bdc8da5abfa3e1/crates/storage/provider/src/providers/mod.rs#L40-L46 

these get updated from the consensus engine (on FCU) and are read from rpc

wdyt @Rjected @rakita

### CODE REVIEW

**Comment by mattsse** [comment]
yeah!
I think this should be controlled by the consensus engine which updates the values on FCU and after pipeline ran

**Comment by Rjected** [comment]
>  I think this should be controlled by the consensus engine which updates the values on FCU and after pipeline ran
yeah this makes sense

**Comment by mattsse** [comment]
implemented parts of this, but really need to find better names for this, any suggestions?

**Comment by Rjected** [comment]
Integrated the `BlockchainProvider` into the `BeaconEngine` so we can access the `ChainInfoTracker` from there, and implemented a helper method for initializing the `BlockchainProvider`:
https://github.com/paradigmxyz/reth/pull/2555/files#diff-7c98a55447a7c335e3d070462eddc12566a7eec7a017b76bc0014242bbb661eaR65-R70

**Comment by Rjected** [comment]
Decided to roll in a refactor of providers - let me know if this hard to review / should be extracted into another PR.

First - all `Provider` traits now use `BlockHashOrNumber` instead of `BlockId`.

Next, I renamed `BlockIdProvider` to `BlockNumProvider`.
`BlockId` and `BlockNumberOrTag` related methods are removed and refactored into their own trait:
```rust
/// Client trait for transforming [BlockId].
#[auto_impl::auto_impl(&, Arc)]
pub trait BlockIdProvider: BlockNumProvider + Send + Sync {
    /// Converts the `BlockNumberOrTag` variants to a block number.
    fn convert_block_number(
        &self,
        num: BlockNumberOrTag,
    ) -> Result<Option<reth_primitives::BlockNumber>> {
        let num = match num {
            BlockNumberOrTag::Latest => self.chain_info()?.best_number,
            BlockNumberOrTag::Earliest => 0,
            BlockNumberOrTag::Pending => {
                return self
                    .pending_block_num_hash()
                    .map(|res_opt| res_opt.map(|num_hash| num_hash.number))
            }
            BlockNumberOrTag::Number(num) => num,
            BlockNumberOrTag::Finalized => return self.finalized_block_num(),
            BlockNumberOrTag::Safe => return self.safe_block_num(),
        };
        Ok(Some(num))
    }

    /// Get the hash of the block by matching the given id.
    fn block_hash_for_id(&self, block_id: BlockId) -> Result<Option<H256>> {
        match block_id {
            BlockId::Hash(hash) => Ok(Some(hash.into())),
            BlockId::Number(num) => {
                if matches!(num, BlockNumberOrTag::Latest) {
                    return Ok(Some(self.chain_info()?.best_hash))
                }

                if matches!(num, BlockNumberOrTag::Pending) {
                    return self
                        .pending_block_num_hash()
                        .map(|res_opt| res_opt.map(|num_hash| num_hash.hash))
                }

                self.convert_block_number(num)?
                    .map(|num| self.block_hash(num))
                    .transpose()
                    .map(|maybe_hash| maybe_hash.flatten())
            }
        }
    }

    /// Get the number of the block by matching the given id.
    fn block_number_for_id(
        &self,
        block_id: BlockId,
    ) -> Result<Option<reth_primitives::BlockNumber>> {
        match block_id {
            BlockId::Hash(hash) => self.block_number(hash.into()),
            BlockId::Number(num) => self.convert_block_number(num),
        }
    }

    /// Get the current pending block number and hash.
    fn pending_block_num_hash(&self) -> Result<Option<reth_primitives::BlockNumHash>>;

    /// Get the safe block number.
    fn safe_block_num(&self) -> Result<Option<reth_primitives::BlockNumber>>;

    /// Get the finalized block number.
    fn finalized_block_num(&self) -> Result<Option<reth_primitives::BlockNumber>>;
}
```
The `pending_block_num_hash`, `safe_block_num`, and `finalized_block_num` methods are all that are needed to implement the trait. The conversion methods should then work automatically.

**Comment by Rjected** [comment]
Will be splitting this up!

**Comment by Rjected** [comment]
@mattsse mind reviewing? seems like I can't request a review since you opened it

File: crates/storage/provider/src/providers/mod.rs [hunk 1]
```diff
@@ -43,12 +46,29 @@ pub struct BlockchainProvider<DB, Tree> {
...
+    }
+}
+
+impl<DB, Tree> BlockchainProvider<DB, Tree>
+where
+    DB: Database,
+{
+    /// Create a new provider using only the database and the tree, fetching the latest header from
+    /// the database to initialize the provider.
+    pub fn new_from_db(database: ShareableDatabase<DB>, tree: Tree) -> Result<Self> {
```
Comment (mattsse) [inline crates/storage/provider/src/providers/mod.rs:66]: I'd like to flip these:

`new -> Result`
`with_latest(SealedHeader) -> Self`

**Review by mattsse — COMMENTED** [review]
a few nits.

I envisioned the `on_forkchoice_update_received` as a simple notification, that only updates a timestamp. I guess we'd need a getter for the timestamp as well so we can add a monitor task that checks it from time to time.

we should handle all validation inside the tree

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -180,13 +181,14 @@ where
...
     BT: BlockchainTreeEngine + 'static,
+    BlockchainProvider<DB, BT>: CanonChainTracker + Unpin + 'static,
 {
     /// Create a new instance of the [BeaconConsensusEngine].
     pub fn new(
         db: DB,
         task_spawner: TS,
         pipeline: Pipeline<DB>,
-        blockchain_tree: BT,
+        blockchain_provider: BlockchainProvider<DB, BT>,
```
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs:191]: yep, changed to `blockchain`

File: crates/consensus/beacon/src/engine/mod.rs [hunk 3]
```diff
@@ -147,8 +148,8 @@ where
     /// Must always be [Some] unless the state is being reevaluated.
     /// The pipeline is used for historical sync by setting the current forkchoice head.
     pipeline_state: Option<PipelineState<DB>>,
-    /// The blockchain tree used for live sync and reorg tracking.
-    blockchain_tree: BT,
+    /// The type we can use to query both the database and the blockchain tree.
+    blockchain_provider: BlockchainProvider<DB, BT>,
```
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs]: Now just using `BT` which is bounded as a `BlockchainTreeEngine + BlockProvider + CanonChainTracker`

File: crates/storage/provider/src/providers/mod.rs [hunk 4]
```diff
@@ -408,6 +428,55 @@ where
     }
 }
 
+impl<DB, Tree> CanonChainTracker for BlockchainProvider<DB, Tree>
+where
+    DB: Send + Sync,
+    Tree: Send + Sync,
+    Self: BlockProvider,
+{
+    fn on_forkchoice_update_received(&self, update: &ForkchoiceState) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/storage/provider/src/providers/mod.rs]: changed it back, and now it just updates the timestamp

File: crates/consensus/beacon/src/engine/mod.rs [hunk 5]
```diff
@@ -380,6 +394,42 @@ where
         Ok(OnForkChoiceUpdated::valid(status))
     }
 
+    /// Sets the state of the canon chain tracker based on the given forkchoice update. This should
+    /// be called before issuing a VALID forkchoice update.
+    fn update_canon_chain(&self, update: &ForkchoiceState) -> Result<(), BeaconEngineError> {
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:397]: perhaps these checks should happen before the `make_canonical` but we can figure this out later

**Review by mattsse — COMMENTED** [review]
awesome!
this is great

**Review by mattsse — COMMENTED** [review]
awesome!
this is great

**Pull request merged**